### PR TITLE
Support Einstein summation convention

### LINKED ISSUE #183: Implement numpy.einsum
**Is your feature request related to a problem? Please describe.**

Implement numpy.einsum

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

support `einsum` operand(https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.einsum.html) for tensor.

## Related issue number

resolves #183.

### CODE REVIEW

File: mars/tensor/base/einsum.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                chunk = tree_add(op.dtype, einsum_chunks, out_idx, tensor_shape, sparse=op.sparse)
+            out_chunks.append(chunk)
+
+        nsplits = [axes_splits[ax] for ax in output_scripts]
+        new_op = op.copy()
+        return new_op.new_tensors(input_tensors, out_tensor.shape, chunks=out_chunks, nsplits=nsplits)
+
+    @classmethod
+    def execute(cls, ctx, op):
+        inputs = [ctx[c.key] for c in op.inputs]
```
Comment (qinxuye) [inline mars/tensor/base/einsum.py]: Use `as_same_device` and `xp` to handle both CPU and GPU `einsum`.

File: mars/tensor/base/einsum.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        ctx[op.outputs[0].key] = np.einsum(op.subscripts, *inputs, optimize=op.optimize, dtype=op.dtype,
+                                           order=op.order, casting=op.casting)
+
+
+def einsum(subscripts, *operands, dtype=None, order='K', casting='safe', optimize=False):
+    """
+    Evaluates the Einstein summation convention on the operands.
```
Comment (qinxuye) [inline mars/tensor/base/einsum.py]: Just copy numpy's `einsum` docs, and do some modification.

File: mars/tensor/base/einsum.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        out_tensor = op.outputs[0]
+        input_scripts, output_scripts = op.subscripts.split('->')
+        tensor_axes = list(zip(op.inputs, input_scripts.split(',')))
+
+        # rechunk to unify nsplits
+        input_nsplits = defaultdict(list)
+        for (t, axes) in tensor_axes:
+            for splits, ax in zip(t.nsplits, axes):
+                input_nsplits[ax].append(splits)
+        input_tensors = []
```
[the remark this answers is not in the document]
Comment (hekaisheng) [reply mars/tensor/base/einsum.py]: the axis is represented by some letters here, it is a bit different from the logic in  `unify_chunks`:
```
tensor_axes = [(t, range(t.ndim)) if not isinstance(t, tuple) else t  for t in tensors]
```

File: mars/tensor/einsum/core.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            if xp is np:
+                ctx[op.outputs[0].key] = xp.einsum(op.subscripts, *inputs, optimize=op.optimize, dtype=op.dtype,
+                                                   order=op.order, casting=op.casting)
+            else:
+                # Cupy doesn't support `optimize`, `order` and `casting`.
+                ctx[op.outputs[0].key] = xp.einsum(op.subscripts, *inputs, dtype=op.dtype)
+
+
+def einsum(subscripts, *operands, dtype=None, order='K', casting='safe', optimize=False):
+    """
```
Comment (qinxuye) [inline mars/tensor/einsum/core.py]: Just copy the doc from numpy, examples needed as well.

**Comment by qinxuye** [comment]
I cannot see the coveralls as well, could you please paste the coverage also?

File: mars/tensor/einsum/core.py (hunk 4 above)
Comment (qinxuye) [reply mars/tensor/einsum/core.py]: Refer to https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/core/einsumfunc.py#L1005

Comment (hekaisheng) [reply mars/tensor/einsum/core.py]: OK, I will add the part of examples.

**Comment by hekaisheng** [comment]
> I cannot see the coveralls as well, could you please paste the coverage also?

It seems failed to upload to coverall, i will retry.

**Pull request merged**